src/app.rs

```diff
@@ -307,35 +311,26 @@ pub fn load_config<P: AsRef<Path>>(path: P) -> io::Result<Config> {
     Ok(config_data)
 }
 
-#[derive(Clone)]
-pub struct AppState {
-    pub db: Addr<DbExecutor>,
-    pub config: Arc<Config>,
-    pub job_queue: Addr<JobQueue>,
-    pub delta_generator: Addr<DeltaGenerator>,
-}
-
-fn handle_build_repo(req: &HttpRequest<AppState>) -> actix_web::Result<NamedFile> {
-    let tail: String = req.match_info().query("tail")?;
-    let id: String = req.match_info().query("id")?;
-    let state = req.state();
+fn handle_build_repo(config: Data<Arc<Config>>,
+                     req: HttpRequest) -> Result<HttpResponse, actix_web::Error> {
+    let tail = req.match_info().query("tail");
+    let id = req.match_info().query("id");
 
-    // Strip out any "../.." or other unsafe things
-    let relpath = PathBuf::from_param(tail.trim_start_matches('/'))?;
-    // The id won't have slashes, but it could have ".." or some other unsafe thing
-    let safe_id = PathBuf::from_param(&id)?;
-    let path = Path::new(&state.config.build_repo_base).join(&safe_id).join(&relpath);
+    // Actix strips out .. and ., but not multiple slashes
```

Comment:
are you sure about this?
`NamedFile::open` seems to simply open the path without removing potential `..`
See: https://docs.rs/actix-files/0.2.0-alpha.2/src/actix_files/named.rs.html#138-140

Reply:
Ugh, i did some testing and it looked like actix rewrote the path before doing the routing. However it turned out I was just being fooled by curl doing the path rewrites (need to do `--path-as-is`).

It is handled in actix-files here https://github.com/actix/actix-web/blob/master/actix-files/src/lib.rs#L596, but that is all private now. I guess we have to duplicate it.

Reply:
@tilosp I added a copy of the actix-files code in a new commit. Does that look ok to you?